url6a9f553762_Bresiu_socket_io_tgz-pJ8-MainJ8/src/com/github/nkzawa/socketio/parser/Parser.java

```diff
@@ -0,0 +1,272 @@
+                encodeAsBinary(obj, callback);
+            } else {
+                String encoding = encodeAsString(obj);
+                callback.call(new String[] {encoding});
+            }
+        }
+
+        private String encodeAsString(Packet obj) {
+            StringBuilder str = new StringBuilder();
+            boolean nsp = false;
+
+            str.append(obj.type);
+
+            if (BINARY_EVENT == obj.type || BINARY_ACK == obj.type) {
+                str.append(obj.attachments);
+                str.append("-");
+            }
+
+            if (obj.nsp != null && obj.nsp.length() != 0 && !"/".equals(obj.nsp)) {
+                nsp = true;
+                str.append(obj.nsp);
+            }
+
+            if (obj.id >= 0) {
+                if (nsp) {
+                    str.append(",");
+                    nsp = false;
+                }
+                str.append(obj.id);
+            }
+
+            if (obj.data != null) {
+                if (nsp) str.append(",");
+                str.append(obj.data);
+            }
+
+            logger.fine(String.format("encoded %s as %s", obj, str));
+            return str.toString();
+        }
+
+        private void encodeAsBinary(Packet obj, Callback callback) {
+            Binary.DeconstructedPacket deconstruction = Binary.deconstructPacket(obj);
+            String pack = encodeAsString(deconstruction.packet);
+            List<Object> buffers = new ArrayList<Object>(Arrays.asList(deconstruction.buffers));
+
+            buffers.add(0, pack);
+            callback.call(buffers.toArray());
+        }
+
+        public interface Callback {
+
+            public void call(Object[] data);
+        }
+    }
+
+    public static class Decoder extends Emitter {
+
+        public static String EVENT_DECODED = "decoded";
+
+        /*package*/ BinaryReconstructor reconstructor;
+
+        public Decoder() {
+            this.reconstructor = null;
+        }
+
+        public void add(String obj) {
+            Packet packet = decodeString(obj);
+            if (BINARY_EVENT == packet.type || BINARY_ACK == packet.type) {
+                this.reconstructor = new BinaryReconstructor(packet);
+
+                if (this.reconstructor.reconPack.attachments == 0) {
+                    this.emit(EVENT_DECODED, packet);
+                }
+            } else {
+                this.emit(EVENT_DECODED, packet);
+            }
+        }
+
+        public void add(byte[] obj) {
+            if (this.reconstructor == null) {
+                throw new RuntimeException("got binary data when not reconstructing a packet");
+            } else {
+                Packet packet = this.reconstructor.takeBinaryData(obj);
+                if (packet != null) {
+                    this.reconstructor = null;
+                    this.emit(EVENT_DECODED, packet);
+                }
+            }
+        }
+
+        private static Packet decodeString(String str) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 19. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 202-229 into a separate method.